port/osx/omrosbacktrace_impl.c

```diff
@@ -0,0 +1,411 @@
+	return unw_backtrace(addresses->address_array, addresses->capacity);
+}
+
+/**
+ * Provides signal protection for the unw_backtrace call. If the stack walk causes a segfault then we check the output
+ * array to see if we got any frames as we may be able to provide a partial backtrace. We also record that the stack walk
+ * was incomplete so that information is available after the function returns.
+ */
+uintptr_t
+backtrace_sigprotect(OMRPortLibrary *portLibrary, J9PlatformThread *threadInfo, void **address_array, const int capacity)
+{
+	uintptr_t ret = 0;
+	frameData args;
+	args.address_array = address_array;
+	args.capacity = capacity;
+
+	memset(address_array, 0, sizeof(void *) * capacity);
+
+	if (omrthread_self()) {
+		if (0 != portLibrary->sig_protect(portLibrary, protectedBacktrace, &args, handler, NULL, OMRPORT_SIG_FLAG_SIGALLSYNC | OMRPORT_SIG_FLAG_MAY_RETURN, &ret)) {
+			/* check to see if there were any addresses populated */
+			for (ret = 0; ret < args.capacity && NULL != address_array[ret]; ret++);
+
+			threadInfo->error = FAULT_DURING_BACKTRACE;
+		}
+
+		return ret;
+	} else {
+		return unw_backtrace(address_array, capacity);
+	}
+}
+
+/**
+ * This function constructs a backtrace from a CPU context. Generally there are only one or two
+ * values in the context that are actually used to construct the stack but these vary by platform
+ * so arn't detailed here. If no heap is specified then this function will use malloc to allocate
+ * the memory necessary for the stack frame structures which must be freed by the caller.
+ *
+ * @param portLibrary a pointer to an initialized port library
+ * @param threadInfo the thread structure we want  to attach the backtrace to. Must not be NULL.
+ * @param heap a heap from which to allocate any necessary memory. If NULL malloc is used instead.
+ * @param signalInfo a platform signal context. If not null the context held in threadInfo is replaced
+ *  	  with the signal context before the backtrace is generated.
+ *
+ * @return the number of frames in the backtrace.
+ */
+uintptr_t
+omrintrospect_backtrace_thread_raw(OMRPortLibrary *portLibrary, J9PlatformThread *threadInfo, J9Heap *heap, void *signalInfo)
+{
+	void *addresses[50];
+	J9PlatformStackFrame **nextFrame = NULL;
+	J9PlatformStackFrame *junkFrames = NULL;
+	J9PlatformStackFrame *prevFrame = NULL;
+	OMRUnixSignalInfo *sigInfo = (OMRUnixSignalInfo *)signalInfo;
+	int i = 0;
+	int discard = 0;
+	int ret = 0;
+	const char *regName = NULL;
+	void **faultingAddress = 0;
+
+	if (NULL == threadInfo) {
+		return 0;
+	}
+
+	/* if we've been passed a port library wrapped signal, then extract info from there */
+	if (NULL != sigInfo) {
+		/* get the faulting address so we can discard frames that are part of the signal handling */
+		infoForControl(portLibrary, sigInfo, 0, &regName, (void **)&faultingAddress);
+	}
+
+	ret = backtrace_sigprotect(portLibrary, threadInfo, addresses, sizeof(addresses) / sizeof(void *));
```

Comment:
Expressing the final argument as `sizeof(addresses) / sizeof(addresses[0])` would be safer.